File: apps/agent-runtime/cocola_agent_runtime/objstore.py
"""Object-store fetcher: pulls attachment bytes on the model's behalf.

Attachment delivery is always push (ADR-0017 P1a): the gateway pre-uploads every
file to the object store (source of truth) and hands agent-runtime either the
inline bytes (small files) or just the object key (large files). This module is
the "given a key, fetch the bytes" side used to materialize the key-only ones
before they are provisioned into ./uploads/.

The gateway owns uploads; agent-runtime only reads. `Fetcher` is a tiny Protocol
so the servicer depends on an abstraction (real MinIO in prod, a fake in tests)
rather than the minio SDK directly -- the same composition-root pattern the rest
of the runtime uses.

Configuration is env-driven (COCOLA_MINIO_*), mirroring the gateway. Object
storage is required by the production composition root so attachments,
artifacts and Skill bundles cannot silently disappear.
"""

from __future__ import annotations

import io
import os
from typing import Protocol

from cocola_common import get_logger

log = get_logger("cocola.agent-runtime.objstore")


class Fetcher(Protocol):
    """Fetches attachments/Skill bundles and publishes output artifacts."""

    def get(self, key: str) -> bytes: ...
    def put(self, key: str, data: bytes, mime: str) -> None: ...


class MinioFetcher:
    """minio-SDK-backed Fetcher. Reads one object fully into memory (attachments
    are size-capped upstream, so a full read is acceptable)."""

    def __init__(self, client, bucket: str) -> None:
        self._client = client
        self._bucket = bucket

    def get(self, key: str) -> bytes:
        resp = None
        try:
            resp = self._client.get_object(self._bucket, key)
            return resp.read()
        finally:
            if resp is not None:
                resp.close()
                resp.release_conn()

    def put(self, key: str, data: bytes, mime: str) -> None:
        self._client.put_object(
            self._bucket,
            key,
            io.BytesIO(data),
            length=len(data),
            content_type=mime or "application/octet-stream",
        )


def fetcher_from_env() -> Fetcher:
    """Build the required MinioFetcher from COCOLA_MINIO_* env.

    Secret key honours the "_FILE" indirection (ADR-0008): if
    COCOLA_MINIO_SECRET_KEY_FILE points at a readable file, its contents (minus a
    trailing newline) are used; otherwise COCOLA_MINIO_SECRET_KEY applies.
    """
    endpoint = os.getenv("COCOLA_MINIO_ENDPOINT", "").strip()
    bucket = os.getenv("COCOLA_MINIO_BUCKET", "").strip()
    access_key = os.getenv("COCOLA_MINIO_ACCESS_KEY", "").strip()
    secret_key = _secret_from_env("COCOLA_MINIO_SECRET_KEY")
    required = {
        "COCOLA_MINIO_ENDPOINT": endpoint,
        "COCOLA_MINIO_ACCESS_KEY": access_key,
        "COCOLA_MINIO_SECRET_KEY": secret_key,
        "COCOLA_MINIO_BUCKET": bucket,
    }
    missing = [name for name, value in required.items() if not value]
    if missing:
        raise RuntimeError(f"{', '.join(missing)} required for object storage")

    # Imported lazily so the dependency is only needed when object storage is
    # actually configured (keeps zero-config local boots import-light).
    from minio import Minio

    secure = os.getenv("COCOLA_MINIO_USE_SSL", "") == "1"

    client = Minio(endpoint, access_key=access_key, secret_key=secret_key, secure=secure)
    log.info("attachment object-store fetcher enabled", bucket=bucket, endpoint=endpoint)
    return MinioFetcher(client, bucket)


def _secret_from_env(name: str) -> str:
    path = os.getenv(name + "_FILE", "").strip()
    if path:
        try:
            with open(path, encoding="utf-8") as fh:
                return fh.read().rstrip("\r\n")
        except OSError as exc:
            raise RuntimeError(f"{name}_FILE is unreadable: {path}") from exc
    return os.getenv(name, "")

File: apps/gateway/internal/objstore/objstore.go
// Package objstore is the gateway's thin client to the S3-compatible object
// store (MinIO in dev, any S3 API in prod) that is the source of truth for
// user-uploaded attachments (ADR-0017 P1a).
//
// The gateway owns uploads: on each /v1/chat it PutObjects every attachment,
// then splits by a configurable size threshold — small files are pushed inline
// into the sandbox, large files carry only their object key and are pulled by
// agent-runtime on the model's behalf.
//
// Store is a narrow interface so the HTTP layer depends on an abstraction (real
// MinIO in prod, a fake in tests) rather than the minio-go stubs directly.
//
// Configuration is env-driven (COCOLA_MINIO_*). Object storage is a required
// part of the production chat path: attachments and durable session state must
// not depend on whether a payload happens to fit in an inline gRPC frame.
package objstore

import (
	"bytes"
	"context"
	"fmt"
	"os"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"

	"github.com/cocola-project/cocola/packages/go-common/config"
)

// Store is the object-storage seam the gateway depends on. Put is the upload
// side of the source-of-truth write; Get is used by tests and health probes
// (agent-runtime does its own Get in Python). Health lets startup/liveness
// verify the bucket is reachable before serving traffic.
type Store interface {
	Put(ctx context.Context, key string, data []byte, mime string) error
	Get(ctx context.Context, key string) ([]byte, error)
	Health(ctx context.Context) error
}

// Config is the resolved object-store configuration. Endpoint is host:port
// without scheme (minio-go takes UseSSL separately). Bucket is the single
// bucket cocola writes to (auto-created by minio-init in dev).
type Config struct {
	Endpoint  string
	AccessKey string
	SecretKey string
	Bucket    string
	UseSSL    bool
}

// ConfigFromEnv reads the COCOLA_MINIO_* environment. The secret key honours
// the "_FILE" indirection (ADR-0008) so a Vault Agent sidecar can render it to
// disk without the app taking a Vault dependency.
func ConfigFromEnv() Config {
	return Config{
		Endpoint:  os.Getenv("COCOLA_MINIO_ENDPOINT"),
		AccessKey: os.Getenv("COCOLA_MINIO_ACCESS_KEY"),
		SecretKey: config.SecretFromEnv("COCOLA_MINIO_SECRET_KEY"),
		Bucket:    os.Getenv("COCOLA_MINIO_BUCKET"),
		UseSSL:    os.Getenv("COCOLA_MINIO_USE_SSL") == "1",
	}
}

func (c Config) Validate() error {
	switch {
	case c.Endpoint == "":
		return fmt.Errorf("objstore: COCOLA_MINIO_ENDPOINT is required")
	case c.AccessKey == "":
		return fmt.Errorf("objstore: COCOLA_MINIO_ACCESS_KEY is required")
	case c.SecretKey == "":
		return fmt.Errorf("objstore: COCOLA_MINIO_SECRET_KEY is required")
	case c.Bucket == "":
		return fmt.Errorf("objstore: COCOLA_MINIO_BUCKET is required")
	default:
		return nil
	}
}

// Client is the minio-go-backed Store.
type Client struct {
	mc     *minio.Client
	bucket string
}

// compile-time assertion that Client satisfies Store.
var _ Store = (*Client)(nil)

// New constructs a MinIO client from cfg. It does not perform network I/O
// (minio-go connects lazily); call Health to verify reachability. Returns an
// error only on malformed configuration.
func New(cfg Config) (*Client, error) {
	if err := cfg.Validate(); err != nil {
		return nil, err
	}
	mc, err := minio.New(cfg.Endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(cfg.AccessKey, cfg.SecretKey, ""),
		Secure: cfg.UseSSL,
	})
	if err != nil {
		return nil, fmt.Errorf("objstore: new client: %w", err)
	}
	return &Client{mc: mc, bucket: cfg.Bucket}, nil
}

// Put uploads data under key with the given content type. It overwrites any
// existing object at that key (keys are uuid-prefixed, so collisions are not
// expected in practice).
func (c *Client) Put(ctx context.Context, key string, data []byte, mime string) error {
	opts := minio.PutObjectOptions{}
	if mime != "" {
		opts.ContentType = mime
	}
	_, err := c.mc.PutObject(ctx, c.bucket, key, bytes.NewReader(data), int64(len(data)), opts)
	if err != nil {
		return fmt.Errorf("objstore: put %q: %w", key, err)
	}
	return nil
}

// Get fetches the whole object at key into memory. Attachments are size-capped
// upstream, so a full read is acceptable here.
func (c *Client) Get(ctx context.Context, key string) ([]byte, error) {
	obj, err := c.mc.GetObject(ctx, c.bucket, key, minio.GetObjectOptions{})
	if err != nil {
		return nil, fmt.Errorf("objstore: get %q: %w", key, err)
	}
	defer func() { _ = obj.Close() }()
	var buf bytes.Buffer
	if _, err := buf.ReadFrom(obj); err != nil {
		return nil, fmt.Errorf("objstore: read %q: %w", key, err)
	}
	return buf.Bytes(), nil
}

// Health verifies the configured bucket exists and is reachable. Used by
// startup wiring and liveness probes.
func (c *Client) Health(ctx context.Context) error {
	ok, err := c.mc.BucketExists(ctx, c.bucket)
	if err != nil {
		return fmt.Errorf("objstore: health: %w", err)
	}
	if !ok {
		return fmt.Errorf("objstore: bucket %q does not exist", c.bucket)
	}
	return nil
}
